srml/balances/src/lib.rs

```diff
@@ -0,0 +1,1210 @@
+	/// has been reduced to zero.
+	///
+	/// Gives a chance to clean up resources associated with the given account.
+	type OnFreeBalanceZero: OnFreeBalanceZero<Self::AccountId>;
+
+	/// Handler for when a new account is created.
+	type OnNewAccount: OnNewAccount<Self::AccountId>;
+
+	/// Handler for the unbalanced reduction when taking fees associated with balance
+	/// transfer (which may also include account creation).
+	type TransferPayment: OnUnbalanced<NegativeImbalance<Self, I>>;
+
+	/// Handler for the unbalanced reduction when removing a dust account.
+	type DustRemoval: OnUnbalanced<NegativeImbalance<Self, I>>;
+
+	/// The overarching event type.
+	type Event: From<Event<Self, I>> + Into<<Self as system::Trait>::Event>;
+
+	/// The minimum amount required to keep an account open.
+	type ExistentialDeposit: Get<Self::Balance>;
+
+	/// The fee required to make a transfer.
+	type TransferFee: Get<Self::Balance>;
+
+	/// The fee required to create an account.
+	type CreationFee: Get<Self::Balance>;
+}
+
+impl<T: Trait<I>, I: Instance> Subtrait<I> for T {
+	type Balance = T::Balance;
+	type OnFreeBalanceZero = T::OnFreeBalanceZero;
+	type OnNewAccount = T::OnNewAccount;
+	type ExistentialDeposit = T::ExistentialDeposit;
+	type TransferFee = T::TransferFee;
+	type CreationFee = T::CreationFee;
+}
+
+decl_event!(
+	pub enum Event<T, I: Instance = DefaultInstance> where
+		<T as system::Trait>::AccountId,
+		<T as Trait<I>>::Balance
+	{
+		/// A new account was created.
+		NewAccount(AccountId, Balance),
+		/// An account was reaped.
+		ReapedAccount(AccountId),
+		/// Transfer succeeded (from, to, value, fees).
+		Transfer(AccountId, AccountId, Balance, Balance),
+	}
+);
+
+/// Struct to encode the vesting schedule of an individual account.
+#[derive(Encode, Decode, Copy, Clone, PartialEq, Eq, RuntimeDebug)]
+pub struct VestingSchedule<Balance, BlockNumber> {
+	/// Locked amount at genesis.
+	pub locked: Balance,
+	/// Amount that gets unlocked every block after `starting_block`.
+	pub per_block: Balance,
+	/// Starting block for unlocking(vesting).
+	pub starting_block: BlockNumber,
+}
+
+impl<Balance: SimpleArithmetic + Copy, BlockNumber: SimpleArithmetic + Copy> VestingSchedule<Balance, BlockNumber> {
+	/// Amount locked at block `n`.
+	pub fn locked_at(&self, n: BlockNumber) -> Balance
+	where
+		Balance: From<BlockNumber>,
+	{
+		// Number of blocks that count toward vesting
+		// Saturating to 0 when n < starting_block
+		let vested_block_count = n.saturating_sub(self.starting_block);
+		// Return amount that is still locked in vesting
+		if let Some(x) = Balance::from(vested_block_count).checked_mul(&self.per_block) {
+			self.locked.max(x) - x
+		} else {
+			Zero::zero()
+		}
+	}
+}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
+pub struct BalanceLock<Balance> {
+	pub id: LockIdentifier,
+	pub amount: Balance,
+	pub until: TimeStamp,
```

Comment:
Can this be a generic Type, support both TimeStamp and BlockNumber?

Reply:
Changed.